File: src/compiler.rs
use crate::schema::CodeNode;
use std::fs;
use std::io::Write;

pub struct ContextCompiler;

impl ContextCompiler {
    pub fn compile(
        focal_target: &str,
        nodes: &[CodeNode],
        resolution: &str,
        direction: &str,
        radius: u8,
        out_filepath: &str,
    ) -> std::io::Result<()> {
        let mut file = fs::File::create(out_filepath)?;

        writeln!(file, "# Dynamic Architectural Context Compilation")?;
        writeln!(file, "Generated deterministically by Celial Graph Engine.\n")?;
        writeln!(file, "## Architectural Context Bounds")?;
        writeln!(file, "- **Focal Target**: `{}`", focal_target)?;
        writeln!(file, "- **Search Direction**: `{}`", direction)?;
        writeln!(file, "- **Traversed Radius**: `{}` hops", radius)?;
        writeln!(file, "- **Resolution Profile**: `{}`", resolution)?;
        writeln!(file, "- **Total Resolved Nodes**: `{}`\n", nodes.len())?;

        writeln!(file, "---")?;
        writeln!(file, "## 1. Focal Core Target Node")?;

        if let Some(core_node) = nodes.iter().find(|n| n.id == focal_target) {
            Self::write_node_details(&mut file, core_node, resolution)?;
        } else {
            writeln!(file, "_Error: Focal node was not compiled in the graph state database._")?;
        }

        writeln!(file, "---")?;
        writeln!(file, "## 2. Traversed Subgraph Neighbors")?;

        let mut sibling_index = 0;
        for node in nodes {
            if node.id == focal_target {
                continue;
            }
            sibling_index += 1;
            writeln!(file, "### Neighbor #{}: `{}`", sibling_index, node.id)?;
            let node_res = if resolution == "full" && radius == 1 { "full" } else { "signature" };
            Self::write_node_details(&mut file, node, node_res)?;
        }

        Ok(())
    }

    fn write_node_details<W: Write>(
        writer: &mut W,
        node: &CodeNode,
        resolution: &str,
    ) -> std::io::Result<()> {
        writeln!(writer, "- **Symbol ID**: `{}`", node.id)?;
        writeln!(writer, "- **Filepath**: `{}`", node.filepath)?;
        writeln!(writer, "- **AST Type**: `{:?}`", node.node_type)?;
        writeln!(
            writer,
            "- **Semantic Facets**: `[Layer: {:?}, Role: {:?}, Pattern: {:?}, Purity: {:?}, Barrier: {}]`",
            node.semantics.layer, node.semantics.role, node.semantics.pattern, node.semantics.purity, node.semantics.purity_barrier
        )?;
        
        let summary_text = node.ai_summary.as_deref().unwrap_or("No cached AI summary available.");
        writeln!(writer, "- **AI Summary**: _{}_", summary_text)?;

        // Expose both previous and current code ranges to facilitate agent's delta summarization loop [5]
        if let Some(ref old_code) = node.previous_code {
            writeln!(writer, "\n#### Outdated Code Block (Prior to modification):")?;
            let code_block_lang = if node.filepath.ends_with(".rs") { "rust" } else { "dart" };
            writeln!(writer, "```{}", code_block_lang)?;
            writeln!(writer, "{}", old_code)?;
            writeln!(writer, "```")?;
        }

        if resolution == "full" {
            writeln!(writer, "\n#### Active Code Block:")?;
            let code_block_lang = if node.filepath.ends_with(".rs") { "rust" } else { "dart" };
            writeln!(writer, "```{}", code_block_lang)?;
            writeln!(writer, "{}", node.raw_code)?;
            writeln!(writer, "```\n")?;
        } else {
            writeln!(writer, "\n_Signature resolution mode: active code block suppressed._\n")?;
        }

        Ok(())
    }
}

File: src/compiler.py
import json
import dataclasses
from pathlib import Path
from typing import List, Optional
from schema import CodeNode, OntologyConfig


class ContextCompiler:

    @staticmethod
    def compile(focal_target: str, nodes: List[CodeNode], resolution: str,
                direction: str, radius: int, out_filepath: str,
                ontology: OntologyConfig) -> None:
        """
        Generates a deterministic markdown-formatted architectural context file
        from the traversed sub-graph.
        """
        out_path = Path(out_filepath)

        with open(out_path, "w", encoding="utf-8") as file:
            file.write("# Dynamic Architectural Context Compilation\n")
            file.write(
                "Generated deterministically by Celial Graph Engine.\n\n")

            file.write(
                "## 0. Active System Ontology Configuration (Boundary Constraints)\n"
            )
            file.write("```json\n")
            # Convert dataclass hierarchy cleanly to dictionary for serialization
            ontology_dict = dataclasses.asdict(ontology)
            file.write(json.dumps(ontology_dict, indent=2) + "\n")
            file.write("```\n\n")

            file.write("## 1. Architectural Context Bounds\n")
            file.write(f"- **Focal Target**: `{focal_target}`\n")
            file.write(f"- **Search Direction**: `{direction}`\n")
            file.write(f"- **Traversed Radius**: `{radius}` hops\n")
            file.write(f"- **Resolution Profile**: `{resolution}`\n")
            file.write(f"- **Total Resolved Nodes**: `{len(nodes)}`\n\n")

            file.write("---\n")
            file.write("## 2. Focal Core Target Node\n")

            # Emulating Rust's iterator slice search path
            core_node = next((n for n in nodes if n.id == focal_target), None)
            if core_node is not None:
                ContextCompiler._write_node_details(file, core_node,
                                                    resolution)
            else:
                file.write(
                    "_Error: Focal node was not compiled in the graph state database._\n"
                )

            file.write("---\n")
            file.write("## 3. Traversed Subgraph Neighbors\n\n")

            sibling_index = 0
            for node in nodes:
                if node.id == focal_target:
                    continue
                sibling_index += 1
                file.write(f"### Neighbor #{sibling_index}: `{node.id}`\n")

                node_res = "full" if resolution == "full" else "signature"
                ContextCompiler._write_node_details(file, node, node_res)

    @staticmethod
    def _write_node_details(writer, node: CodeNode, resolution: str) -> None:
        """Internal helper to stream structured code node metadata into the markdown generator."""
        writer.write(f"- **Symbol ID**: `{node.id}`\n")
        writer.write(f"- **Filepath**: `{node.filepath}`\n")
        # Accessing value property of the String-backed Enum
        writer.write(f"- **AST Type**: `'{node.node_type.value}'`\n")

        sem = node.semantics
        writer.write(
            f"- **Semantic Facets**: `[Layer: {sem.layer}, Role: {sem.role}, "
            f"Pattern: {sem.pattern}, Purity: {sem.purity}, Barrier: {sem.purity_barrier}]`\n"
        )

        summary_text = node.ai_summary if node.ai_summary is not None else "No cached AI summary available."
        writer.write(f"- **AI Summary**: _{summary_text}_\n")

        if resolution == "full":
            writer.write("\n#### Active Code Block:\n")
            code_block_lang = "rust" if node.filepath.endswith(
                ".rs") else "dart"
            writer.write(f"```{code_block_lang}\n")
            writer.write(f"{node.raw_code}\n")
            writer.write("```\n\n")
        else:
            writer.write(
                "\n_Signature resolution mode: active code block suppressed._\n\n"
            )
